### PR TITLE
Columnar: Fix ALTER TABLE ... ADD COLUMN.

### PR DESCRIPTION
**Pull request by pykello**
The relation we get in the read stage of ALTER TABLE is the altered relation, so we shouldn't use its tuple descriptor to read the contents of current data file. We should instead use the tuple descriptor of slot passed to columnar_getnextslot.

When adding a column with default value, postgres creates a transient relation, then reads from the old relation and writes to the new relation adding the new column, and then swaps relfilenodes between two relations. At the stage which reads from the old relation we need to pass the tuple descriptor before the ALTER TABLE. But unfortunately alters the table before the "read", so in `init_columnar_read_state` the `relation` doesn't contain the correct tuple descriptor for reading. Instead the tuple descriptor passed as `tupdesc` which comes from the first tuple read is correct one.

This is similar to https://github.com/citusdata/citus/pull/4320

Fixes https://github.com/citusdata/citus/issues/4602
Fixes https://github.com/citusdata/citus/issues/4601

### CODE REVIEW

File: src/test/regress/expected/am_alter.out [hunk 1]
```diff
@@ -171,3 +173,58 @@ ALTER TABLE test_alter_table ALTER COLUMN j TYPE int;
...
+  7 | t
+  8 | t
+  9 | t
+ 10 | t
+(10 rows)
+
+-- https://github.com/citusdata/citus/issues/4601
+CREATE TABLE itest13 (a int) using columnar;
+INSERT INTO itest13 VALUES (1), (2), (3);
+ALTER TABLE itest13 ADD COLUMN c int GENERATED BY DEFAULT AS IDENTITY;
```
Comment (marcocitus) [inline src/test/regress/expected/am_alter.out:213]: should we also have tests for regular generated columns? (e.g. an expression)

**Comment by thanodnl** [comment]
The description of the PR makes a lot of sense, I fail to see how that translates to the changes we are making. The fix requires a refactor.

Are both calls to `RelationColumnList` affected by the issue?

File: src/test/regress/expected/am_alter.out [hunk 2]
```diff
@@ -171,3 +173,58 @@ ALTER TABLE test_alter_table ALTER COLUMN j TYPE int;
 ALTER TABLE test_alter_table ALTER COLUMN k TYPE varchar(20);
 ALTER TABLE test_alter_table ALTER COLUMN k TYPE text;
 DROP TABLE test_alter_table;
+-- https://github.com/citusdata/citus/issues/4602
+create domain str_domain as text not null;
+create table domain_test (a int, b int) using columnar;
+insert into domain_test values (1, 2);
+insert into domain_test values (1, 2);
+alter table domain_test add column c str_domain;
+ERROR:  domain str_domain does not allow null values
```
Comment (jeff-davis) [inline src/test/regress/expected/am_alter.out:182]: Please add a comment indicating this is a negative test

**Review by jeff-davis — APPROVED** [review]

**Comment by pykello** [comment]
> The description of the PR makes a lot of sense, I fail to see how that translates to the changes we are making. The fix requires a refactor.
> 
> Are both calls to `RelationColumnList` affected by the issue?

No, only the `init_columnar_read_state` one is affected. I added some comments to description to explain this more. This requires refactoring on postres side. If postgres passed the relation with correct tuple descriptor to read to us, we didn't need to do the state initialization when reading the first tuple and we could do it in `columnar_beginscan`.

**Pull request merged**